cpp/src/arrow/csv/reader.cc

```diff
@@ -500,6 +510,7 @@ class ReaderMixin {
 
   // Number of columns in the CSV file
   int32_t num_csv_cols_ = -1;
+  int64_t num_rows_;
```

Comment:
nit: maybe something more like `last_row_number_` or `num_seen_rows_` and a comment noting `-1` means we aren't tracking the row count?